### PR TITLE
Refactor Python MatrixRead

### PR DESCRIPTION
**Pull request by jigold**
- used by read_matrix_table, range_matrix_table, import_vcf, import_bgen

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/Parser.scala [hunk 1]
```diff
@@ -936,13 +937,13 @@ object IRParser {
         val rowExpr = ir_value_expr(newEnv)(it)
         MatrixAggregateRowsByKey(child, entryExpr, rowExpr)
       case "MatrixRead" =>
-        val typ = opt(it, matrix_type_expr)
+        val requestedType = opt(it, matrix_type_expr)
         val dropCols = boolean_literal(it)
         val dropRows = boolean_literal(it)
         val readerStr = string_literal(it)
-        implicit val formats = MatrixReader.formats
+        implicit val formats = MatrixReader.formats + new MatrixBGENReaderSerializer(env)
```
Comment (catoverdrive) [inline hail/src/main/scala/is/hail/expr/ir/Parser.scala]: can we add this to MatrixReader.formats?

File: hail/src/main/scala/is/hail/HailContext.scala [hunk 2]
```diff
@@ -677,7 +677,7 @@ class HailContext private(val sc: SparkContext,
         ReferenceGenome.addReference(referenceGenome)
         true
       } else false
-    }
+    } // Needed for tests
```
Comment (catoverdrive) [inline hail/src/main/scala/is/hail/HailContext.scala]: is it too much of a lift to move this into testutils?

File: hail/src/main/scala/is/hail/expr/ir/Parser.scala (hunk 1 above)
Comment (jigold) [reply hail/src/main/scala/is/hail/expr/ir/Parser.scala]: I don't think so. I tried to parameterize the formats by the env, but it's also used by Pretty where there is no environment.

Comment (catoverdrive) [reply hail/src/main/scala/is/hail/expr/ir/Parser.scala]: oh, I see. I missed the part where the BGEN reader depends on the env.

**Pull request merged**